### PR TITLE
Vector clocks

### PR DESCRIPTION
**Pull request by Pat-Lafon**
Provides ordering to a list of operations that are waiting to be applied

### CODE REVIEW

File: src/main/java/VectorClock.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    public BigInteger get(String key) {
+        BigInteger val = vectorclock.get(key);
+        if (val == null) {
+            return BigInteger.ZERO; // Provide 0 as a default value
+        }
+        return val;
+    }
+
+    public void maxClock(VectorClock c) {
```
Comment (mpmilano) [inline src/main/java/VectorClock.java:45]: name suggestion -> advanceTo

File: src/main/java/VectorClock.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        return true;
+    }
+
+    @Override
+    public int compareTo(VectorClock c) {
+        if (this.lessthan(c))
+            return -1;
+        if (c.lessthan(this))
+            return 1; // greater than
+        return 0;
```
Comment (mpmilano) [inline src/main/java/VectorClock.java:68]:
```java
else { /*lexically order clocks:  */
let keyset = this.keys ∪ c.keys;  
sort(keyset); 
for (key : keyset) { 
    if (this[key] < c[key]) {return -1} 
    else if (this[key] > c[key] ) { return 1} 
}
return 0;
}
```

File: src/main/java/VectorClockBackend.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        }
+
+        CRDTMapEntry maptriple = new CRDTMapEntry(new_crdt_object, new HashSet<Integer>(), new_crdt_effect_buffer);
+        ObjectTable.put(new_key, maptriple);
+        OtpErlangObject[] emptypayload = new OtpErlangObject[2];
+        emptypayload[0] = new_key;
+        emptypayload[1] = new OtpErlangBinary(crdt_object);
+        OtpErlangTuple new_snapshot = new OtpErlangTuple(emptypayload);
+        return new_snapshot;
+    }
```
Comment (mpmilano) [inline src/main/java/VectorClockBackend.java:134]: Your snapshots must contain all updates which are more recent than the "global min time"

File: src/main/java/VectorClock.java (hunk 1 above)
Comment (Pat-Lafon) [reply src/main/java/VectorClock.java]: UpdateClock

File: src/main/java/VectorClock.java (hunk 2 above)
Comment (Pat-Lafon) [reply src/main/java/VectorClock.java:82]: Done

File: src/main/java/VectorClockBackend.java (hunk 3 above)
Comment (Pat-Lafon) [reply src/main/java/VectorClockBackend.java]: Whoops

File: src/main/java/CRDTMapEntry.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import java.io.Serializable;
+import java.util.TreeSet;
+
+public class CRDTMapEntry implements Serializable {
```
[the remark this answers is not in the document]
Comment (Pat-Lafon) [reply src/main/java/CRDTMapEntry.java:4]: Makes sense

File: src/main/java/Snapshot.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import java.io.Serializable;
+import java.util.TreeSet;
+
+public class Snapshot implements Serializable {
+    private static final long serialVersionUID = 12L;
+    public CRDT object;
+    public TreeSet<GenericEffect> effectbuffer;
+
+    public Snapshot(CRDT Object, TreeSet<GenericEffect> sortedEffectSet) {
+        this.object = Object;
```
Comment (mpmilano) [inline src/main/java/Snapshot.java:10]: maybe avoid using popular class names as variables man.

File: src/main/java/VectorClockBackend.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            throw new NoSuchObjectException();
+        }
+
+        // Apply any effects that have passed the clock time to the new copy of the
+        // object
+        TreeSet<GenericEffect> crdt_effect_buffer = mapentry.effectbuffer;
+        TreeSet<GenericEffect> new_crdt_effect_buffer = new TreeSet<GenericEffect>();
+        for (GenericEffect e : crdt_effect_buffer) {
+            // Use a correct compare based on types
+            if (e.time.lessthan(this.GlobalClockTime)) {
```
Comment (mpmilano) [inline src/main/java/VectorClockBackend.java:98]: e.time.lessthan(this.GlobalClockTime) || (0 == e.time.compareTo(this.GlobalClockTime))

File: src/main/java/VectorClockBackend.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    public OtpErlangBinary update(OtpErlangBinary JavaObjectId, OtpErlangBinary binary) throws NoSuchObjectException {
+        if (!ObjectTable.containsKey(JavaObjectId)) {
+            try {
+                CRDT crdt_object = ((CRDTEffect) binary.getObject()).crdt;
+                Snapshot mapentry = new Snapshot(crdt_object, new TreeSet<GenericEffect>());
+                ObjectTable.put(JavaObjectId, mapentry);
+            } catch (ClassCastException e) {
+                // We don't have the object and we have been given an update
+                // so we need to request the object from antidote and try again
+                assert (binary.getObject().getClass() == GenericFunction.class);
```
[the remark this answers is not in the document]
Comment (Pat-Lafon) [reply src/main/java/VectorClockBackend.java]: Yep, changed

File: src/main/java/VectorClockBackend.java (hunk 6 above)
Comment (Pat-Lafon) [reply src/main/java/VectorClockBackend.java]: Sounds good

**Pull request merged**